src/ngraph/node.hpp

```diff
@@ -596,139 +580,65 @@ namespace ngraph
-        const element::Type& get_element_type() const
-        {
-            return m_node->get_input_element_type(m_index);
-        }
-        /// \return The shape of the input referred to by this input handle.
-        const Shape& get_shape() const { return m_node->get_input_shape(m_index); }
-        /// \return The partial shape of the input referred to by this input handle.
-        const PartialShape& get_partial_shape() const
-        {
-            return m_node->get_input_partial_shape(m_index);
-        }
-        /// \return A handle to the output that is connected to this input.
-        Output<Node> get_source_output() const;
-        /// \return A reference to the tensor descriptor for this input.
-        descriptor::Tensor& get_tensor() const
-        {
-            return m_node->m_inputs.at(m_index).get_output().get_tensor();
-        }
-        /// \return A shared pointer to the tensor descriptor for this input.
-        std::shared_ptr<descriptor::Tensor> get_tensor_ptr() const
-        {
-            return m_node->m_inputs.at(m_index).get_output().get_tensor_ptr();
-        }
-        /// \return true if this input is relevant to its node's output shapes; else false.
-        bool get_is_relevant_to_shapes() const
-        {
-            return m_node->m_inputs.at(m_index).get_is_relevant_to_shape();
-        }
-        /// \return true if this input is relevant to its node's output values; else false.
-        bool get_is_relevant_to_values() const
-        {
-            return m_node->m_inputs.at(m_index).get_is_relevant_to_value();
-        }
-
-        bool operator==(const Input& other) const
-        {
-            return m_node == other.m_node && m_index == other.m_index;
-        }
-        bool operator!=(const Input& other) const { return !(*this == other); }
-        bool operator<(const Input& other) const
-        {
-            return m_node < other.m_node || (m_node == other.m_node && m_index < other.m_index);
-        }
-        bool operator>(const Input& other) const
-        {
-            return m_node > other.m_node || (m_node == other.m_node && m_index > other.m_index);
-        }
-        bool operator<=(const Input& other) const { return !(*this > other); }
-        bool operator>=(const Input& other) const { return !(*this < other); }
-    private:
-        const Node* const m_node;
+        const Node* m_node;
         const size_t m_index;
     };
 
     /// \brief A handle for one of a node's outputs.
-    template <>
-    class NGRAPH_API Output<Node>
+    class NGRAPH_API NodeOutput
     {
     public:
-        /// \brief Constructs a Output.
+        /// \brief Constructs a NodeOutput.
         /// \param node A pointer to the node for the output handle.
         /// \param index The index of the output.
-        Output(Node* node, size_t index)
-            : m_node(node->shared_from_this())
+        NodeOutput(const Node* node, size_t index)
```

Comment:
`const_node.output(0).get_node()` loses constness.

Reply:
Yes, but it does exactly what it needs to do. That is the intent of the const_cast. This is what we need to make this work without the ugly templates